syncapi/consumers/roomserver.go

```diff
@@ -154,41 +154,71 @@ func (s *OutputRoomEventConsumer) onNewRoomEvent(
 	ctx context.Context, msg api.OutputNewRoomEvent,
 ) error {
 	ev := msg.Event
-
 	addsStateEvents := []*gomatrixserverlib.HeaderedEvent{}
-	foundEventIDs := map[string]bool{}
-	if len(msg.AddsStateEventIDs) > 0 {
-		for _, eventID := range msg.AddsStateEventIDs {
-			foundEventIDs[eventID] = false
+
+	// Work out the list of events we need to find out about. Either
+	// they will be the event supplied in the request, we will find it
+	// in the sync API database or we'll need to ask the roomserver.
+	knownEventIDs := make(map[string]bool, len(msg.AddsStateEventIDs))
+	for _, eventID := range msg.AddsStateEventIDs {
+		knownEventIDs[eventID] = eventID == ev.EventID()
+	}
+
+	// Work out which events we want to look up in the sync API database.
+	// At this stage the only event that should be excluded is the event
+	// supplied in the request, if it appears in the adds_state_event_ids.
+	missingEventIDs := make([]string, 0, len(msg.AddsStateEventIDs))
+	for eventID, known := range knownEventIDs {
+		if !known {
+			missingEventIDs = append(missingEventIDs, eventID)
 		}
-		foundEvents, err := s.db.Events(ctx, msg.AddsStateEventIDs)
+	}
+
+	// Look the events up in the database. If we know them, add them into
+	// the set of adds state events.
+	if len(missingEventIDs) > 0 {
+		alreadyKnown, err := s.db.Events(ctx, msg.AddsStateEventIDs)
 		if err != nil {
 			return fmt.Errorf("s.db.Events: %w", err)
 		}
-		for _, event := range foundEvents {
-			foundEventIDs[event.EventID()] = true
+		for _, knownEvent := range alreadyKnown {
+			knownEventIDs[knownEvent.EventID()] = true
+			addsStateEvents = append(addsStateEvents, knownEvent)
```

Comment:
This still isn't done for the initial event if there are no missing events.

Reply:
I would not add incrementally to `addStateEvents` like this. We know that the events we want to add all have keys in the unhelpfully named `knownEventIDs`, we just need to loop that and append rather than do it twice (and forget as in this case).